From: Julian Andres Klode <jak@debian.org>
Date: Fri, 25 Jan 2019 21:04:17 +0100
Subject: kernel 5.0: Replace do_gettimeofday() with ktime_get_real()

do_gettimeofday() was removed in kernel 5.0, a direct replacement
seems to be ktime_get_real() and friends. ktime_get_real() seems
to be a nice replacement, and even makes the code easier.

Bug-Ubuntu: https://bugs.launchpad.net/bugs/1813063
---
 driver/ntoskernel.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/driver/ntoskernel.c b/driver/ntoskernel.c
index 156c688..a99490b 100644
--- a/driver/ntoskernel.c
+++ b/driver/ntoskernel.c
@@ -2498,7 +2498,11 @@ struct worker_init_struct {
 
 int ntoskernel_init(void)
 {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0)
 	struct timeval now;
+#else
+	ktime_t now;
+#endif
 
 	spin_lock_init(&dispatcher_lock);
 	spin_lock_init(&ntoskernel_lock);
@@ -2518,11 +2522,18 @@ int ntoskernel_init(void)
 	INIT_WORK(&ntos_work, ntos_work_worker);
 	wrap_timer_slist.next = NULL;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0)
 	do_gettimeofday(&now);
 	wrap_ticks_to_boot = TICKS_1601_TO_1970;
 	wrap_ticks_to_boot += (u64)now.tv_sec * TICKSPERSEC;
 	wrap_ticks_to_boot += now.tv_usec * 10;
 	wrap_ticks_to_boot -= jiffies * TICKSPERJIFFY;
+#else
+	now = ktime_get_real();
+	wrap_ticks_to_boot = TICKS_1601_TO_1970;
+	wrap_ticks_to_boot += ktime_to_us(now) * 10;
+	wrap_ticks_to_boot -= jiffies * TICKSPERJIFFY;
+#endif
 	TRACE2("%llu", wrap_ticks_to_boot);
 
 	cpu_count = num_online_cpus();
